### PR TITLE
[FR] Implement Hourly Account Balance updator & start implementing Cu…

### PR DESCRIPTION
**Pull request by anerhan**
…stomers

### CODE REVIEW

File: app/assets/javascripts/profile.js [hunk 1]
```diff
@@ -16,17 +17,18 @@
...
         event.preventDefault();
+      }).on('change', function(e){
+        App.profile.dobIsValid = App.profile.dobRegexp.test($(e.target).val())
+        App.profile.setFormState();
       });
     },
 
     isValid: function(){
+      console.log('DOB', App.profile.dobIsValid);
+      console.log('SSN', App.profile.ssnIsValid);
```
Comment (agm1988) [inline app/assets/javascripts/profile.js]: don't forget to remove debug

File: app/models/customer.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
+class Customer < ApplicationRecord
+  # INFO: https://docsv2.dwolla.com/#customers
+  STATUSES = %w[unverified verified retry document suspended deactivated].freeze
+  TYPES = %w[unverified personal business receive-only].freeze
+
+  enum status: STATUSES.each_with_object({}) { |val, hash| hash[val.to_sym] = val }
+
+  attribute :status, :string, default: STATUSES[0]
```
Comment (agm1988) [inline app/models/customer.rb]: default option should be set (at db level) in migration

File: app/models/customer.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  has_many :funding_sources, dependent: :destroy
+
+  delegate :email, to: :profile, allow_nil: true
+  delegate :user, to: :profile, allow_nil: true
+
+  def dwolla_registered?
+    uid.present? && link.present?
+  end
+
+  def business_name
```
Comment (agm1988) [inline app/models/customer.rb]: this always raises an error?

File: app/models/funding_source.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
+class FundingSource < ApplicationRecord
+  STATUSES = %w[pending processed failed].freeze
+
+  attribute :status, :string, default: STATUSES[0]
+
+  enum status: STATUSES.each_with_object({}) { |val, hash| hash[val.to_sym] = val }
```
Comment (agm1988) [inline app/models/funding_source.rb]: why you are using both `attribute` and `enum` for status?

File: app/processors/plaid/fetch_account_info.rb [hunk 5]
```diff
@@ -24,17 +24,18 @@ def process!
...
         # TODO: Maybe need handle get plaid identity info for creating customer
         plaid_identity = PlaidService.account_identity(account.plaid_token, account.uid)
+
         if plaid_identity
           account.plaid_identity = plaid_identity
           if account.is_a?(SourceAccount)
             user_profile = account.user.profile || account.user.build_profile
             account.plaid_identity.profile_info.each do |k, v|
-              user_profile.send(:"#{k}=", v) if user_profile.send(:"#{k}")&.blank?
+              user_profile.send(:"#{k}=", v) if user_profile.send(:"#{k}").blank?
```
Comment (agm1988) [inline app/processors/plaid/fetch_account_info.rb]: use `public_send`

File: app/processors/plaid/fetch_balances.rb [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    class FetchBalances
+      attr_reader :service
+
+      def initialize(opts = {}); end
+
+      def process!
+        InvestSet.activated.includes(:source_account, :invest_account).find_each(batch_size: 30).each do |is|
+          next unless is.ready?
+
+          InvestSet.transaction do
```
Comment (agm1988) [inline app/processors/plaid/fetch_balances.rb]: if `PlaidService` fails, would it's error raise ActiveRecord Rollback?

File: app/processors/register_invest_set_customers.rb [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+      raise(ArgumentError, 'Option :id required!') if opts[:id].blank?
+
+      @invest_set = InvestSet.includes(:source_account, :invest_account).find(opts[:id])
+    end
+
+    def process!
+      return unless invest_set.ready?
+
+      InvestSet.transaction do
+        Processors::Dwolla::RegisterCustomer.new(account: invest_set.source_account).process!
```
Comment (agm1988) [inline app/processors/register_invest_set_customers.rb]: same here failure error raises rollback?

File: app/services/plaid_service.rb [hunk 8]
```diff
@@ -31,14 +31,14 @@ def account_data(opts = {})
...
+  def find_by(opts = {})
+    raise(ArgumentError, 'Opts conditions required!') if opts.empty?
 
-    raise(ArgumentError, "Option :type should be in range [#{ACCOUNT_TYPES.join(',')}]!") unless type.in?(ACCOUNT_TYPES)
-
-    accounts.detect { |acc| acc.uid == opts[:uid] }[:balances][type]
+    accounts.detect do |acc|
+      opts.map { |k, v| acc.send(:"#{k}") == v }.uniq.exclude?(false)
+    end
+  rescue StandardError
```
Comment (agm1988) [inline app/services/plaid_service.rb]: better to rescue specific errors, but unless here is no custom error - should be good to go

File: app/views/layouts/blocks/_navbar.html.erb [hunk 9]
```diff
@@ -18,6 +18,11 @@
             <li class="user-menu__dropdown__item">
               <%= link_to 'Dashboard', dashboard_index_path %>
             </li>
+            <% if current_user.admin? %>
```
Comment (agm1988) [inline app/views/layouts/blocks/_navbar.html.erb]: `current_user` nil is handled here?

File: app/views/layouts/rails_admin/pjax.html.haml [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+%title= "#{@abstract_model.try(:pretty_name) || @page_name} | #{[_get_plugin_name[0] || 'Rails', _get_plugin_name[1] || 'Admin'].join(' ')}"
+%p
+- flash && flash.each do |key, value|
+  .alert.alert-dismissible{class: flash_alert_class(key)}
+    %button.close{type: 'button', :'data-dismiss' => "alert"} &times;
+    = value
+= breadcrumb
+%ul.nav.nav-tabs
+  = menu_for((@abstract_model ? (@object.try(:persisted?) ? :member : :collection) : :root), @abstract_model, @object)
```
Comment (agm1988) [inline app/views/layouts/rails_admin/pjax.html.haml]: "hard" to understand check, it's from rails admin source?

File: spec/factories/customer.rb [hunk 11]
```diff
@@ -0,0 +1,12 @@
+FactoryBot.define do
+  factory :customer do
+    link { File.join('https://api-sandbox.dwolla.com/customers', SecureRandom.uuid) }
+    customer_type { 'unverified' }
+    status { 'unverified' }
+    trait :with_profile do
+      after(:build) do |customer|
+        customer.profile = create(:profile)
```
Comment (agm1988) [inline spec/factories/customer.rb]: maybe better to use `build`?

File: spec/factories/funding_source.rb [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+    trait :failed do
+      after(:build) do |funding_source|
+        funding_source.status = 'failed'
+      end
+    end
+
+    trait :with_customer do
+      after(:build) do |funding_source|
+        funding_source.customer = create(:customer, :with_profile)
```
Comment (agm1988) [inline spec/factories/funding_source.rb]: maybe better to use build?

File: spec/factories/profile.rb [hunk 13]
```diff
@@ -3,5 +3,7 @@
     user
     first_name { 'Some name' } # TODO: add faker
     last_name { 'Some last name' }
+    dob { Date.current }
```
Comment (agm1988) [inline spec/factories/profile.rb]: this will result in invalid dob for dwolla is it ok?

File: app/models/funding_source.rb [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
+  before_save do
+    self.uid ||= link.split('/').last if link.present?
+  end
+
+  validates :customer_id, :account_id, :status, presence: true
+
+  belongs_to :customer
```
Comment (agm1988) [inline app/models/funding_source.rb]: add `inverse_of` for such relations

File: app/models/customer.rb (hunk 3 above)
Comment (anerhan) [reply app/models/customer.rb]: Sure if you call this method under Customer instance object. It's Ruby interface pattern. This method should be implemented in all children classes like InvestCustomer and SourceCustomer

File: spec/factories/profile.rb (hunk 13 above)
Comment (anerhan) [reply spec/factories/profile.rb]: Ok. Beacuse dob before send will be transformed to dwolla format

File: app/models/customer.rb (hunk 3 above)
Comment (agm1988) [reply app/models/customer.rb]: aha, understood. Error like: "Should be implemented" would be more explicit. ok :+1:

File: spec/factories/profile.rb (hunk 13 above)
Comment (agm1988) [reply spec/factories/profile.rb]: I meant Date.current < 21 years old

File: app/views/layouts/rails_admin/pjax.html.haml (hunk 10 above)
Comment (anerhan) [reply app/views/layouts/rails_admin/pjax.html.haml]: Sure, from admin)

File: spec/factories/profile.rb (hunk 13 above)
Comment (anerhan) [reply spec/factories/profile.rb]: Sure You're write

File: app/models/funding_source.rb (hunk 14 above)
Comment (anerhan) [reply app/models/funding_source.rb]: Will implement later

**Pull request merged**